backend/common/src/test/java/ai/verta/modeldb/common/nats/JetstreamConnectorTest.java

```diff
@@ -0,0 +1,158 @@
+import static org.awaitility.Awaitility.await;
+
+import io.nats.client.*;
+import io.nats.client.api.ConsumerConfiguration;
+import io.nats.client.api.ConsumerInfo;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import lombok.extern.log4j.Log4j2;
+import np.com.madanpokharel.embed.nats.EmbeddedNatsConfig;
+import np.com.madanpokharel.embed.nats.EmbeddedNatsServer;
+import np.com.madanpokharel.embed.nats.NatsServerConfig;
+import np.com.madanpokharel.embed.nats.NatsVersion;
+import org.junit.jupiter.api.*;
+
+@Log4j2
+class JetstreamConnectorTest {
+  public static EmbeddedNatsServer embeddedNatsServer;
+  private JetstreamConnector connector;
+
+  @BeforeAll
+  static void beforeAll() {
+    EmbeddedNatsConfig natsConfig;
+    try {
+      natsConfig =
+          new EmbeddedNatsConfig.Builder()
+              .withNatsServerConfig(
+                  new NatsServerConfig.Builder()
+                      .withEnableJetStream(
+                          Files.createTempDirectory("jetstream").toAbsolutePath().toString())
+                      .withConfigParam("--no_advertise", "")
+                      .withRandomPort()
+                      .withHost("localhost")
+                      .withNatsVersion(NatsVersion.LATEST)
+                      .build())
+              .build();
+      embeddedNatsServer = new EmbeddedNatsServer(natsConfig);
+      embeddedNatsServer.startServer();
+      await().until(() -> embeddedNatsServer.isServerRunning());
+      Runtime.getRuntime().addShutdownHook(new Thread(() -> embeddedNatsServer.stopServer()));
+    } catch (Exception e) {
+      log.error("Could not start embedded NATS server", e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  @BeforeEach
+  void setUp() {
+    JetstreamConfig config =
+        new JetstreamConfig(
+            true, embeddedNatsServer.getRunningHost(), embeddedNatsServer.getRunningPort(), 1);
+    connector = new JetstreamConnector(config, () -> List.of("cheese", "wine"));
+  }
+
+  @AfterEach
+  void tearDown() {
+    if (connector != null) {
+      connector.close();
+    }
+  }
+
+  @AfterAll
+  static void afterAll() {
+    if (embeddedNatsServer != null) {
+      embeddedNatsServer.stopServer();
+    }
+  }
+
+  @Test
+  void publishAndSubscribe() throws Exception {
+    AtomicReference<String> message = new AtomicReference<>();
+    connector.subscribeToStream(
+        "cheese",
+        "cheese.>",
+        msg -> {
+          msg.ack();
+          message.set(new String(msg.getData()));
+        },
+        "testDurable",
+        "testGroup");
+
+    JetStream cheeseStream = connector.getJetStream("cheese");
+    cheeseStream.publish("cheese.create", "Swiss".getBytes(StandardCharsets.UTF_8));
+
+    await().untilAsserted(() -> assertThat(message.get()).isEqualTo("Swiss"));
```

Comment:
I was wondering if we could capture a exception that occurs within the event handler. I tried to do a test in the registry for this but I couldn't find a way to capture the exception and do some assertion on it. I was trying with the "messageHandler" object which was the only one available I saw. Do you know if there's a way to do it with this EmbeddedNatsServer project?

Reply:
What specific case are you thinking of? When our handler throws an exception, or some other exception?

Reply:
When the handler throws exception. For example that parse message method that can throw a exception of not being able to parse, I couldn't catch it. I think it is because it is running async...

Reply:
hm. that's probably a test of the handler, rather than a test of this code, I believe. All we can do here is provide a handler that is implemented however we need to test that it's being called appropriately.